File: driver-service/src/main/java/com/dulfinne/taxi/driverservice/exception/GlobalExceptionHandler.java
package com.dulfinne.taxi.driverservice.exception;

import com.dulfinne.taxi.driverservice.util.ExceptionKeys;
import lombok.RequiredArgsConstructor;
import org.springframework.context.MessageSource;
import org.springframework.context.i18n.LocaleContextHolder;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.HashMap;
import java.util.Map;

@RestControllerAdvice
@RequiredArgsConstructor
public class GlobalExceptionHandler {

  private final MessageSource validationMessageSource;
  private final MessageSource exceptionMessageSource;

  @ExceptionHandler(MethodArgumentNotValidException.class)
  public ResponseEntity<Map<String, String>> handleValidationExceptions(
      MethodArgumentNotValidException ex) {
    Map<String, String> errors = new HashMap<>();
    for (FieldError fieldError : ex.getBindingResult().getFieldErrors()) {
      String fieldName = fieldError.getField();
      String errorMessage = fieldError.getDefaultMessage();

      errors.put(
          fieldName,
          validationMessageSource.getMessage(errorMessage, null, LocaleContextHolder.getLocale()));
    }
    return new ResponseEntity<>(errors, HttpStatus.BAD_REQUEST);
  }

  @ExceptionHandler(EntityNotFoundException.class)
  public ResponseEntity<ErrorResponse> handleEntityNotFoundException(EntityNotFoundException ex) {
    String message =
        exceptionMessageSource.getMessage(
            ex.getMessageKey(), ex.getParams(), LocaleContextHolder.getLocale());
    ErrorResponse errorResponse = new ErrorResponse(HttpStatus.NOT_FOUND, message);
    return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorResponse);
  }

  @ExceptionHandler(EntityAlreadyExistsException.class)
  public ResponseEntity<ErrorResponse> handleEntityAlreadyExistsException(
      EntityAlreadyExistsException ex) {

    String message =
        exceptionMessageSource.getMessage(
            ex.getMessageKey(), ex.getParams(), LocaleContextHolder.getLocale());
    ErrorResponse errorResponse = new ErrorResponse(HttpStatus.CONFLICT, message);
    return ResponseEntity.status(HttpStatus.CONFLICT).body(errorResponse);
  }

  @ExceptionHandler(Exception.class)
  public ResponseEntity<ErrorResponse> handleGlobalException(Exception ex) {
    String message =
        exceptionMessageSource.getMessage(
            ExceptionKeys.UNKNOWN_ERROR, null, LocaleContextHolder.getLocale());

    ErrorResponse errorResponse = new ErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR, message);
    return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
  }
}

File: ride-service/src/main/kotlin/com/dulfinne/taxi/rideservice/exception/GlobalExceptionHandler.kt
package com.dulfinne.taxi.rideservice.exception

import com.dulfinne.taxi.rideservice.util.ExceptionKeys
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.MessageSource
import org.springframework.context.i18n.LocaleContextHolder
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler(
    private val validationMessageSource: MessageSource,
    private val exceptionMessageSource: MessageSource
) {
    private val log: Logger = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)

    @ExceptionHandler(ClientException::class)
    fun handleClientException(ex: ClientException): ResponseEntity<ErrorResponse> {
        val status = HttpStatus.valueOf(ex.code)
        val errorResponse = ErrorResponse(status, ex.info)

        log.warn("Feign client exception. Handling. Message: ${ex.info}")
        return ResponseEntity.status(status).body(errorResponse)
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleHttpMessageNotReadableException(ex: HttpMessageNotReadableException): ResponseEntity<ErrorResponse> {
        val errorResponse = ErrorResponse(HttpStatus.BAD_REQUEST, ex.message ?: "JSON parse error")

        log.warn("HttpMessageNotReadableException. Handling.", ex)
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationExceptions(ex: MethodArgumentNotValidException): ResponseEntity<Map<String, String>> {
        val errors = ex.bindingResult.fieldErrors.associate { fieldError ->
            val fieldName = fieldError.field
            val errorMessage = fieldError.defaultMessage ?: "Invalid value"
            fieldName to validationMessageSource.getMessage(errorMessage, null, LocaleContextHolder.getLocale())
        }

        log.warn("Validation exception. Handling. Message: $errors")
        return ResponseEntity(errors, HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(EntityNotFoundException::class)
    fun handleEntityNotFoundException(ex: EntityNotFoundException): ResponseEntity<ErrorResponse> {
        val message = exceptionMessageSource.getMessage(
            ex.messageKey, ex.params ?: emptyArray(), LocaleContextHolder.getLocale()
        )
        val errorResponse = ErrorResponse(HttpStatus.NOT_FOUND, message)
        log.info("Entity not found. Handling. Message: $message")
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorResponse)
    }

    @ExceptionHandler(ActionNotAllowedException::class)
    fun handleActionNotAllowedException(ex: ActionNotAllowedException): ResponseEntity<ErrorResponse> {
        val message = exceptionMessageSource.getMessage(
            ex.messageKey, ex.params ?: emptyArray(), LocaleContextHolder.getLocale()
        )
        val errorResponse = ErrorResponse(HttpStatus.CONFLICT, message)

        log.info("Action not allowed. Handling. Message: $message")
        return ResponseEntity.status(HttpStatus.CONFLICT).body(errorResponse)
    }

    @ExceptionHandler(ServiceNotAvailableException::class)
    fun handleServiceNotAvailableException(ex: ServiceNotAvailableException): ResponseEntity<ErrorResponse> {
        val message = exceptionMessageSource.getMessage(
            ex.messageKey, null, LocaleContextHolder.getLocale()
        )
        val errorResponse = ErrorResponse(HttpStatus.SERVICE_UNAVAILABLE, message)
        log.info("Service not available. Handling. Message: $message")
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorResponse)
    }

    @ExceptionHandler(Exception::class)
    fun handleGlobalException(ex: Exception): ResponseEntity<ErrorResponse> {
        val message =
            exceptionMessageSource.getMessage(ExceptionKeys.UNKNOWN_ERROR, null, LocaleContextHolder.getLocale())

        val errorResponse = ErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR, message)
        log.error("Unknown exception. Handling.", ex);
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse)
    }
}